src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs

```diff
@@ -184,22 +186,69 @@ private async Task<IReadOnlyList<string>> GetResourceIdsForLogAnalyticsAsync(IRe
+                // The workspaces associated with each resource have to be retrieved.
                 var workspacesList = new List<ResourceIdentifier>();
-                foreach (string subscriptionId in resources.Select(resource => resource.SubscriptionId).Distinct(StringComparer.CurrentCultureIgnoreCase))
+
+                foreach (ResourceIdentifier resource in resources)
                 {
-                    // Try to get the workspaces list from the cache, and if it isn't there, use the Azure Resource Manager client
-                    IList<ResourceIdentifier> subscriptionWorkspaces;
-                    if (!this.subscriptionIdToWorkspaces.TryGetValue(subscriptionId, out subscriptionWorkspaces))
+                    // If the resource is a KubernetesService cluster then get its associated workspace directly. Otherwise all workspaces in its subscription must be retrieved.
+                    if (resource.ResourceType == ResourceType.KubernetesService)
                     {
-                        subscriptionWorkspaces = await this.azureResourceManagerClient.GetAllResourcesInSubscriptionAsync(subscriptionId, new[] { ResourceType.LogAnalytics }, cancellationToken);
-                        this.subscriptionIdToWorkspaces[subscriptionId] = subscriptionWorkspaces;
+                        // Retrieve the specific workspace the target Kubernetes cluster belongs to.
+                        if (!this.aksClusterIdToWorkspaces.TryGetValue(resource.ToString(), out ResourceIdentifier workspace))
+                        {
+                            // Try to get the workspaces from the cache, and if it isn't there, use the Azure Resource Manager client
+                            ResourceProperties resourceProperties = await this.azureResourceManagerClient.GetResourcePropertiesAsync(resource, cancellationToken);
+
+                            // (get values from Json in a case-insensitive way)
+                            if (((resourceProperties.Properties?.GetValue("addonProfiles", StringComparison.CurrentCultureIgnoreCase) as
+                                JObject)?.GetValue("omsagent", StringComparison.CurrentCultureIgnoreCase) as
+                                JObject)?.GetValue("enabled", StringComparison.CurrentCultureIgnoreCase)?.ToObject<bool>() ?? false)
+                            {
+                                string idstring = (((resourceProperties.Properties.GetValue("addonProfiles", StringComparison.CurrentCultureIgnoreCase) as
+                                    JObject)?.GetValue("omsagent", StringComparison.CurrentCultureIgnoreCase) as
+                                    JObject)?.GetValue("config", StringComparison.CurrentCultureIgnoreCase) as
+                                    JObject)?.GetValue("loganalyticsworkspaceresourceid", StringComparison.CurrentCultureIgnoreCase).ToString();
+

```

Comment:
It works - but it's a pretty cumbersome way to do it... Shorter would be:
resourceProperties.Properties?["addonProfiles"]?["omsagent"]?["config"]?["loganalyticsworkspaceresourceid"]?.ToString()

Reply:
I think, cleaner way is to convert the JObject to string and convert the entire string to lower case and then convert the lower case string into JObject.  Here is the sample snippet

var resourcePropertiesString = JsonConvert.SerializeObject(resourceProperties);
var lowerCaseResourcePropertiesString = resourcePropertiesString.ToLower();
JObject resultResourceProperties = JObject.Parse(lowerCaseResourcePropertiesString);